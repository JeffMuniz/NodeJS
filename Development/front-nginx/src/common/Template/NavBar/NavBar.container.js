import React, { Component } from "react";
import { arrayOf, bool, func, shape, number, string } from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import jwtDecode from "jwt-decode";
import { getNotifications } from "src/api/notifications/notifications";
import store from "src/redux/configureStore";
import DateManager from "src/modules/DateManager/DateManager";
import { getEnvironment } from "src/modules/UrlManager/UrlManager";
import { resetAuth as resetAuthAction } from "src/redux/modules/session/actions/session";
import { OpenModal as openModalAction } from "src/redux/modules/modal/actions/modal";
import { Routes, WebPaths } from "src/routes/consts";
import { dropDownMenu, notificationsTypes } from "@enums";

import NavBar from "./NavBar";
import GroupSelect from "../GroupSelect/GroupSelect.container";

export class NavBarContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isDropDownOpen: false,
      dropDownHasBorder: false,
      isDropDownNotification: false,
      isContractDownload: false,
      companySegment: null,
      groupIdControl: false,
    };
  }

  async componentDidMount() {
    await this.getJwt();
    this.updateDropdownBorderState();

    const { history } = this.props;

    history.listen(() => {
      this.updateDropdownBorderState();
    });
  }

  async componentDidUpdate() {
    await this.getJwt();

    if (getEnvironment() !== "production") {
      const { groupIdControl } = this.state;
      const {
        user: {
          profile: {
            data: { id: userId },
          },
        },
        selectedCompanyTree: {
          selectedGroup: { id: groupId },
        },
      } = store.getState();
      if (userId && groupId && !groupIdControl) {
        this.updateGroupIdControl();
        await this.getNotification({ userId, groupId });
      }
    }
  }

  getJwt = async () => {
    const { accessToken } = store.getState().session;
    if (
      DateManager(jwtDecode(accessToken).exp, "X")
        .add(4, "hours")
        .isBefore(new Date())
    ) {
      this.handleLogout();
    }
  };

  getNotification = async ({ userId, groupId }) => {
    if (userId && groupId) {
      const { data } = await getNotifications({
        userId,
        groupId,
        notificationType: notificationsTypes.ATUALIZACAO_CONTRATO,
      });
      if (data.segmentoEmpresa)
        this.setState({
          isContractDownload: true,
          companySegment: data.segmentoEmpresa,
        });
    }
    return this.state;
  };

  updateGroupIdControl = () => {
    this.setState({ groupIdControl: true });
  };

  handleLogout = () => {
    const {
      resetAuth,
      history: { push },
    } = this.props;

    this.toggleDropDown();

    resetAuth();

    push(WebPaths[Routes.LOGIN]);
  };

  handleGroupModal = () => {
    const { isDropDownOpen } = this.state;
    const { openModal } = this.props;

    if (isDropDownOpen) this.toggleDropDown();
    openModal({
      children: <GroupSelect />,
    });
  };

  toggleDropDown = () => {
    const { isDropDownOpen } = this.state;
    return this.setState({ isDropDownOpen: !isDropDownOpen });
  };

  toggleNotification = () => {
    const { isDropDownNotification, isDropDownOpen } = this.state;
    if (isDropDownOpen) {
      return this.setState({ isDropDownNotification: false });
    }
    return this.setState({ isDropDownNotification: !isDropDownNotification });
  };

  updateDropdownBorderState = () => {
    const {
      history: {
        location: { pathname },
      },
    } = this.props;
    const { dropDownHasBorder: prevBorderState } = this.state;

    const dropDownHasBorder = dropDownMenu.some(
      link => WebPaths[link.route] === pathname,
    );

    if (prevBorderState !== dropDownHasBorder) {
      this.setState({ dropDownHasBorder });
    }
  };

  render() {
    const { isAuthenticated, groupList, deliveryType } = this.props;
    const {
      isDropDownOpen,
      dropDownHasBorder,
      isContractDownload,
      isDropDownNotification,
      companySegment,
    } = this.state;
    const showChangeGroup = groupList && groupList.length > 1;
    const { length: groupListLength } = groupList;

    return (
      <NavBar
        deliveryType={deliveryType}
        isAuthenticated={isAuthenticated}
        isDropDownOpen={isDropDownOpen}
        isDropDownNotification={isDropDownNotification}
        isContractDownload={isContractDownload}
        companySegment={companySegment}
        toggleDropDown={this.toggleDropDown}
        toggleNotification={this.toggleNotification}
        handleLogout={this.handleLogout}
        groupListLength={groupListLength}
        showChangeGroup={showChangeGroup}
        handleGroupModal={this.handleGroupModal}
        dropDownHasBorder={dropDownHasBorder}
      />
    );
  }
}

export const mapStateToProps = ({
  user: { isAuthenticated },
  group: {
    groupState: { groupList },
  },
  selectedCompanyTree: {
    selectedGroup: {
      params: { deliveryType },
    },
  },
}) => ({
  isAuthenticated,
  groupList,
  deliveryType,
});

export const mapDispatchToProps = {
  resetAuth: resetAuthAction,
  openModal: openModalAction,
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(NavBarContainer),
);

NavBarContainer.propTypes = {
  isAuthenticated: bool,
  resetAuth: func.isRequired,
  groupList: arrayOf(
    shape({
      id: number,
      idPermission: number,
      name: string,
    }),
  ),
  openModal: func.isRequired,
  deliveryType: string,
};

NavBarContainer.defaultProps = {
  isAuthenticated: false,
  groupList: [],
  deliveryType: "",
};
