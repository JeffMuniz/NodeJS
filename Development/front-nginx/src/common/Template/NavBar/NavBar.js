import React, { Component } from "react";
import { bool, func, number, string } from "prop-types";
import { navBarMenu } from "@enums";
import { If } from "@utils";
import { getPDFDocument } from "src/api/files/files";
import { getmacPortal } from "src/modules/UrlManager/UrlManager";
import { snakeCase } from "lodash";

import { WebPaths } from "src/routes/consts";

import { Icon } from "@common";
import {
  Nav,
  NavContainer,
  NavCol,
  NavRow,
  NavLinkStyled,
  activeLink,
  Avatar,
  DropDownCol,
  DropDownColNotification,
  CircleNotificationRed,
  ContainerNotification,
  ContainerDrop,
  DropDownWrapper,
  DownloadContract,
} from "./NavBar.styles";

import DropDown from "./DropDown/DropDown";

class Navbar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isDownloadSuccess: false,
    };
    this.triggerRef = React.createRef();
  }

  componentWillUnmount() {
    this.setState({
      isDownloadSuccess: false,
    });
  }

  onClickDownloadContract = companySegment => {
    const urlSegment = this.checkSegmentCompany(companySegment);
    try {
      const fileName = `${snakeCase(
        `Contrato Atualizado ${companySegment}`,
      )}.pdf`;
      const url = `${getmacPortal()}${urlSegment}`;
      getPDFDocument({ url, fileName });
      // postNotificationContract();
      this.setState({ isDownloadSuccess: true });
    } catch (e) {
      this.setState({ isDownloadSuccess: true });
      // eslint-disable-next-line no-console
      console.log(e);
    }
  };

  checkSegmentCompany = companySegment => {
    const urlE1 = "static/pdf/ContratoServicoRH_RCN_20190110-E1.pdf";
    const urlE2 = "static/pdf/ContratoServicoRH_RCN_20190110-E2.pdf";
    const urlE3 = "static/pdf/ContratoServicoRH_RCN_20190110-E3.pdf";
    if (companySegment.includes("E1")) return urlE1;
    if (companySegment.includes("E2")) return urlE2;
    if (companySegment.includes("E3")) return urlE3;
  };

  render() {
    const {
      isAuthenticated,
      isDropDownOpen,
      isDropDownNotification,
      toggleDropDown,
      handleLogout,
      groupListLength,
      showChangeGroup,
      handleGroupModal,
      deliveryType,
      toggleNotification,
      isContractDownload,
      companySegment,
    } = this.props;

    const { isDownloadSuccess } = this.state;
    const isAuthIsContract = isContractDownload && !!isAuthenticated;

    if (isContractDownload && !!isAuthenticated && !isDownloadSuccess) {
      return (
        <Nav id="nav_navbar">
          <NavContainer>
            <NavRow middle="xs">
              <NavCol xs={1}>
                <Icon name="mac" />
              </NavCol>
              {isAuthenticated && (
                <NavCol xs={8}>
                  {navBarMenu.map(link => (
                    <If test={!link.hidden} key={link.route}>
                      <NavLinkStyled
                        id={`nav_link_${link.route}`}
                        to={WebPaths[link.route]}
                        activeStyle={activeLink}
                      >
                        {link.label}
                      </NavLinkStyled>
                    </If>
                  ))}
                </NavCol>
              )}
              <ContainerNotification xs={3}>
                <CircleNotificationRed onClick={toggleNotification}>
                  1
                </CircleNotificationRed>
                <DropDownColNotification
                  id="nav_link_notification"
                  onClick={toggleNotification}
                >
                  <Icon name="notification" />
                </DropDownColNotification>
                <DropDownCol id="nav_link_userData" onClick={toggleDropDown}>
                  <div ref={this.triggerRef}>
                    <Avatar>
                      <Icon name="person" />
                    </Avatar>
                  </div>
                </DropDownCol>
              </ContainerNotification>
            </NavRow>
            {isAuthenticated && (
              <DropDown
                deliveryType={deliveryType}
                isOpen={isDropDownOpen}
                handleClickOutside={toggleDropDown}
                triggerRef={this.triggerRef}
                handleLogout={handleLogout}
                groupListLength={groupListLength}
                showChangeGroup={showChangeGroup}
                handleGroupModal={handleGroupModal}
              />
            )}
            {isAuthIsContract && !isDownloadSuccess && (
              <DropDownWrapper isOpen={isDropDownNotification}>
                <p>Seu contrato foi atualizado </p>
                <DownloadContract
                  onClick={() => this.onClickDownloadContract(companySegment)}
                >
                  Clique aqui para atualizar
                </DownloadContract>
              </DropDownWrapper>
            )}
          </NavContainer>
        </Nav>
      );
    }
    return (
      <Nav id="nav_navbar">
        <NavContainer>
          <NavRow middle="xs">
            <NavCol xs={1}>
              <Icon name="mac" />
            </NavCol>
            {isAuthenticated && (
              <NavCol xs={9}>
                {navBarMenu.map(link => (
                  <If test={!link.hidden} key={link.route}>
                    <NavLinkStyled
                      id={`nav_link_${link.route}`}
                      to={WebPaths[link.route]}
                      activeStyle={activeLink}
                    >
                      {link.label}
                    </NavLinkStyled>
                  </If>
                ))}
              </NavCol>
            )}
            {isAuthenticated && (
              <ContainerDrop lg={2}>
                <DropDownCol id="nav_link_userData" onClick={toggleDropDown}>
                  <div ref={this.triggerRef}>
                    <Avatar>
                      <Icon name="person" />
                    </Avatar>
                  </div>
                </DropDownCol>
              </ContainerDrop>
            )}
          </NavRow>
          {isAuthenticated && (
            <DropDown
              deliveryType={deliveryType}
              isOpen={isDropDownOpen}
              handleClickOutside={toggleDropDown}
              triggerRef={this.triggerRef}
              handleLogout={handleLogout}
              groupListLength={groupListLength}
              showChangeGroup={showChangeGroup}
              handleGroupModal={handleGroupModal}
            />
          )}
          {isAuthIsContract && !isDownloadSuccess && (
            <DropDownWrapper isOpen={isDropDownNotification}>
              <p>Seu contrato foi atualizado </p>
              <DownloadContract
                onClick={() => this.onClickDownloadContract(companySegment)}
              >
                Clique aqui para atualizar
              </DownloadContract>
            </DropDownWrapper>
          )}
        </NavContainer>
      </Nav>
    );
  }
}

Navbar.propTypes = {
  isAuthenticated: bool.isRequired,
  handleLogout: func.isRequired,
  isContractDownload: bool,
  isDropDownOpen: bool,
  isDropDownNotification: bool,
  toggleDropDown: func,
  toggleNotification: func,
  groupListLength: number,
  showChangeGroup: bool,
  handleGroupModal: func,
  companySegment: string,
  deliveryType: string.isRequired,
};

Navbar.defaultProps = {
  isDropDownOpen: false,
  isDropDownNotification: false,
  isContractDownload: false,
  toggleDropDown: () => undefined,
  toggleNotification: () => undefined,
  groupListLength: 0,
  showChangeGroup: false,
  companySegment: null,
  handleGroupModal: () => null,
};

export default Navbar;
