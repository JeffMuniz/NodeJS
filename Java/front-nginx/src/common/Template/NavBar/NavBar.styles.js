import styled from "styled-components";
import { NavLink } from "react-router-dom";
import { Col, Row } from "react-flexbox-grid";
import { Container } from "@common";
import { navBarGrey, white, red, blue, rgbaWhite } from "src/styles/colors";
import { ifStyle } from "@utils";

const isActive = ifStyle("active");
const isOpen = ifStyle("isOpen");

export const Nav = styled.nav`
  background-color: ${navBarGrey};
  height: 96px;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 4;
`;

export const NavContainer = styled(Container)`
  position: relative;
  background-color: transparent;
`;

export const NavCol = styled(Col)`
  height: 100%;
  align-items: center;
  display: flex;
`;

export const DropDownCol = styled(Col)`
  align-items: center;
  padding-left: 20px;
  box-sizing: border-box;
  border-bottom: 4px solid ${isActive(red, navBarGrey)};
  cursor: pointer;
  display: flex;
  height: 100%;
  justify-content: center;
  text-align: center;
  transition: border 0.25s ease-in-out;
  & > div {
    align-items: center;
    justify-content: center;
    display: flex;
    width: 100%;
    height: 100%;
  }
`;

export const ContainerNotification = styled(Col)`
  display: flex;
  justify-content: flex-end;
`;

export const ContainerDrop = styled(Col)`
  display: flex;
  justify-content: flex-end;
`;

export const DropDownColNotification = styled(Col)`
  align-items: center;
  padding-left: 85px;
  cursor: pointer;
`;

export const DropDownWrapper = styled.div`
  background-color: ${white};
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  display: flex;
  flex-direction: column;
  opacity: ${isOpen("1", "0")};
  padding: 5px 24px 10px 24px;
  position: absolute;
  right: 0;
  transition: opacity 0.25s ease-in-out;
  z-index: 4;
  width: 175px;
  visibility: ${isOpen("visible", "hidden")};
  border-radius: 2px;

  & > p {
    font-size: 12px;
  }

  & > p::before {
    content: "";
    position: absolute;
    width: 0;
    height: 0;
    bottom: 100%;
    left: 97px;
    border: 0.75rem solid transparent;
    border-top: none;
    border-bottom-color: #fff;
    filter: drop-shadow(0 -0.0625rem 0.0625rem rgba(0, 0, 0, 0.1));
  }
`;

export const DownloadContract = styled.a`
  font-size: 13px;
  text-decoration: none;
  cursor: pointer;
  color: ${blue};
`;

export const CircleNotificationRed = styled.span`
  position: absolute;
  width: 30px;
  height: 30px;
  right: 6%;
  background: red;
  border: 1px solid #353638;
  border-radius: 50%;
  top: 8px;
  text-align: center;
  font-size: 18px
  cursor: pointer;
  color: ${white};
`;

export const NavRow = styled(Row)`
  background-color: ${navBarGrey};
  height: 96px;
`;

export const NavLinkStyled = styled(NavLink)`
  align-items: center;
  color: ${white};
  border-bottom: 4px solid ${navBarGrey};
  box-sizing: border-box;
  display: flex;
  height: 100%;
  margin: 0 15px;
  padding: 0 15px;
  text-decoration: none;
  transition: border 0.25s ease-in-out;
`;

export const activeLink = {
  borderBottom: `4px solid ${red}`,
};

export const Avatar = styled.div`
  align-items: center;
  background-color: ${blue};
  border-radius: 30px;
  color: ${rgbaWhite};
  cursor: pointer;
  display: flex;
  font-weight: bold;
  font-size: 20px;
  height: 48px;
  justify-content: center;
  user-select: none;
  width: 48px;
`;
