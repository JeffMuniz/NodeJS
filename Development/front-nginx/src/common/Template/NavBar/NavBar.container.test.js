import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import TestRenderer from "react-test-renderer";
import { NavBarContainer, mapStateToProps } from "./NavBar.container";

jest.mock("./NavBar", () => "NavBar");
jest.mock("../GroupSelect/GroupSelect.container", () => "GroupSelect");
jest.mock("@common", () => ({
  SvgIcon: "SvgIcon",
}));

let defaultProps = null;
let testRenderer = null;
let instance = null;

beforeEach(() => {
  defaultProps = {
    isAuthenticated: false,
    userData: { name: "User Test" },
    resetAuth: jest.fn(),
    history: {
      location: {
        pathname: "/",
      },
      listen: jest.fn(),
    },
    openModal: jest.fn(),
    deliveryType: "HR",
  };
  testRenderer = TestRenderer.create(<NavBarContainer {...defaultProps} />);
  instance = testRenderer.getInstance();
});

describe("NavBar container", () => {
  it("should render it", () => {
    // given
    const renderer = new ShallowRenderer();

    // when
    renderer.render(<NavBarContainer {...defaultProps} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });

  it("should render navbarcontainer with userData empty", () => {
    // given
    const renderer = new ShallowRenderer();
    const props = {
      ...defaultProps,
      isAuthenticated: false,
      userData: {},
      resetAuth: jest.fn(),
      history: {},
      openModal: jest.fn(),
      deliveryType: "HR",
    };

    // when
    renderer.render(<NavBarContainer {...props} />);
    const result = renderer.getRenderOutput();

    // then
    expect(result).toMatchSnapshot();
  });

  it("should toggle isDropDownOpen when toggleDropDown is called", () => {
    // given

    // when
    instance.toggleDropDown();

    // then
    expect(instance.state.isDropDownOpen).toBeTruthy();

    // when
    instance.toggleDropDown();

    // then
    expect(instance.state.isDropDownOpen).not.toBeTruthy();
  });

  it("mapStateToProps should return a object with user from passed state", () => {
    // given
    const state = {
      user: { isAuthenticated: false },
      group: {
        groupState: { groupList: [] },
      },
      selectedCompanyTree: {
        selectedGroup: {
          params: { deliveryType: "HR" },
        },
      },
    };

    // when
    const result = mapStateToProps(state);

    // then
    expect(result).toEqual({
      isAuthenticated: false,
      groupList: [],
      deliveryType: "HR",
    });
  });
});
