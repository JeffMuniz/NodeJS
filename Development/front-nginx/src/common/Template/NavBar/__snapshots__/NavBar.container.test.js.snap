// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar container should render it 1`] = `
<NavBar
  companySegment={null}
  deliveryType="HR"
  dropDownHasBorder={false}
  groupListLength={0}
  handleGroupModal={[Function]}
  handleLogout={[Function]}
  isAuthenticated={false}
  isContractDownload={false}
  isDropDownNotification={false}
  isDropDownOpen={false}
  showChangeGroup={false}
  toggleDropDown={[Function]}
  toggleNotification={[Function]}
/>
`;

exports[`NavBar container should render navbarcontainer with userData empty 1`] = `
<NavBar
  companySegment={null}
  deliveryType="HR"
  dropDownHasBorder={false}
  groupListLength={0}
  handleGroupModal={[Function]}
  handleLogout={[Function]}
  isAuthenticated={false}
  isContractDownload={false}
  isDropDownNotification={false}
  isDropDownOpen={false}
  showChangeGroup={false}
  toggleDropDown={[Function]}
  toggleNotification={[Function]}
/>
`;
